### PR TITLE
Update process should check version and not allow reversion through update

### PR DESCRIPTION
**Pull request by parsharma**
Issue: https://github.com/wildfly-extras/prospero/issues/759

### CODE REVIEW

**Comment by TomasHofman** [comment]
Relevant wildfly-channel issue that I think we would need implemented there in order to complete this functionality:
https://github.com/wildfly-extras/wildfly-channel/issues/292

**Comment by TomasHofman** [comment]
A good next step could be to obtain a list of manifests used by the current installation. I think it should be possible to obtain maven GAVs (or URLs) of the current manifests. If there are any Maven based manifest found in the list, they should be matched against the list of new manifests (which the user want to upgrade to) used by the ChannelSession.

(That list of new manifests is currently simulated by the `findManifestUpdates()` method, hard-coded to only return `org.jbosss.eap.channels:eap-8.0` manifest artifact, because we don't know how to obtain these data.)

**Comment by parsharma** [comment]
Hello @TomasHofman,
I have updated the code where we are getting list of manifests from the current installation.

**Comment by TomasHofman** [comment]
Now can we get the list of manifests currently used by the installation? Maybe from the InstallationMetadata class?

If we can have that, we can try to obtain an intersection from those two lists (manifests groupId:artifactId s that are present in both currently installed manifest and the newly available manifests used by the channelSession).

Once we have that intersection, we can try to compare the versions to see if they are upgrading or downgrading.

**Comment by parsharma** [comment]
@TomasHofman 
I have added the new method compareManifest() where this method compares the currently installed Maven manifests with the available manifest updates from the ChannelSession. It finds the manifests that exist in both the installed manifests and the available updates (based on groupId and artifactId).  For each matching manifest, it compares the version of the installed manifest with the version from the available update. Depending on the comparison, the method outputs whether an upgrade, downgrade, or no change is required.

File: prospero-cli/src/main/java/org/wildfly/prospero/cli/commands/UpdateCommand.java [hunk 1]
```diff
@@ -471,4 +478,49 @@ public static Path detectProsperoInstallationPath() throws ArgumentParsingExcept
...
+
+            if (updateArtifact != null) {
+                // Compare versions
+                String installedVersion = installedManifest.getVersion();
+                String availableVersion = updateArtifact.getVersion();
+
+                if (installedVersion.compareTo(availableVersion) < 0) {
+                    System.out.println("Upgrade available for " + installedManifest.getArtifactId() + ": " + installedVersion + " -> " + availableVersion);
+                } else if (installedVersion.compareTo(availableVersion) > 0) {
+                    System.out.println("Downgrade detected for " + installedManifest.getArtifactId() + ": " + installedVersion + " -> " + availableVersion);
```
Comment (TomasHofman) [inline prospero-cli/src/main/java/org/wildfly/prospero/cli/commands/UpdateCommand.java:518]: I think the comparison logic here should not be an alphabetic comparison, but something that understands semantic versioning.

The problem is that e.g. version "1.11" would be evaluated as lower than "1.2" when using alphabetical ordering, while it should  be higher semantically.

Comment (parsharma) [reply prospero-cli/src/main/java/org/wildfly/prospero/cli/commands/UpdateCommand.java]: Thanks for noticing it @TomasHofman 
I have the code where I am using DefaultArtifactVersion which we compare the versions semantically.

Comment (spyrkob) [reply prospero-cli/src/main/java/org/wildfly/prospero/cli/commands/UpdateCommand.java]: `org.wildfly.channel.version.VersionMatcher` has the correct version compare logic

File: prospero-cli/src/main/java/org/wildfly/prospero/cli/commands/UpdateCommand.java [hunk 2]
```diff
@@ -471,4 +479,49 @@ public static Path detectProsperoInstallationPath() throws ArgumentParsingExcept
...
+                if (update.getGroupId().equals(installedManifest.getGroupId()) &&
+                        update.getArtifactId().equals(installedManifest.getArtifactId())) {
+                    intersectingManifests.add(installedManifest); // Add to intersection list
+                    break;
+                }
+            }
+        }
+
+        // For each intersecting manifest, compare the version
+        for (ManifestVersionRecord.MavenManifest installedManifest : intersectingManifests) {
```
Comment (spyrkob) [inline prospero-cli/src/main/java/org/wildfly/prospero/cli/commands/UpdateCommand.java:499]: you don't need the `intersectingManifest` list. Just iterate over mavenManifests. You're already ignoring ones that are not in `updateArtifacts`.

**Comment by parsharma** [comment]
The CI/CD is failing because we need wildfly-channel upgrade for the API.

**Comment by TomasHofman** [comment]
Yes, we must ask Jeff to release.

**Comment by TomasHofman** [comment]
Never mind, he did and you upgraded already.

**Comment by TomasHofman** [comment]
@spyrkob currently this only checks for downgrade in the `update perform` command. There is the second pathway with `update prepare` / `update apply`. Should the check be made during `prepare` or `apply` stage? If I prepare and update, and then the installation changes, is the prepared update still applicable or do I need to prepare it again?

**Comment by TomasHofman** [comment]
Maybe we could move the check into `UpdateCommand#buildUpdate()` method, which would perform the check both during `update perform` and `update prepare`.

**Comment by spyrkob** [comment]
>  I prepare and update, and then the installation changes, is the prepared update still applicable or do I need to prepare it again?

If the base server changes from the state that the candidate was generated from, that candidate is no longer valid and cannot be applied.

**Pull request closed without merging**